### PR TITLE
Implement test for fsck on the resulting incremental after multiple snapshots

### LINKED ISSUE #254: Implement test for `fsck` on the resulting incremental after multiple snapshots
https://github.com/elastio/elastio-snap/blob/feat/xfs-draft-rebase3/tests/test_incrementals.sh this test should be rewritten on Python.

There is an idea about checksum check for the mounted backup and/or for the files in it.

### PR DESCRIPTION
**Pull request by skypodolsky**
Added python test for multiple update-img usage and some basic data integrity check

Closes #254

### CODE REVIEW

**Review by vsazhenyuk-softheme — APPROVED** [review]

**Review by anelson — APPROVED** [review]

File: tests/devicetestcase.py [hunk 1]
```diff
@@ -30,7 +30,8 @@ def setUpClass(cls):
         if os.getenv('TEST_DEVICES'):
             cls.devices = os.getenv('TEST_DEVICES').split()
             for device in cls.devices:
-                util.dd("/dev/zero", device, util.dev_size_mb(device), bs="1M")
+                cls.size_mb = util.dev_size_mb(device)
```
Comment (e-kov) [inline tests/devicetestcase.py:33]: As I recall, there were 2 cases. 1st case where `TEST_DEVICES` is not set and the tests use 1 loopback device. And 2nd case with usually 2 devices created or attached to machine in advance. Usually they are of the same size, but shouldn't be an array of sizes by device?

File: tests/test_update_image.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env python3
+# SPDX-License-Identifier: GPL-2.0-only
+
+#
+# Copyright (C) 2019 Datto, Inc.
+# Additional contributions by Elastio Software, Inc are Copyright (C) 2020 Elastio Software Inc.
+#
```
Comment (e-kov) [inline tests/test_update_image.py:7]: Old header.

File: tests/test_update_image.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+import elastio_snap
+import util
+from devicetestcase import DeviceTestCase
+
+
+class TestUpdateImage(DeviceTestCase):
+    def setUp(self):
+        self.snap_mount = "/mnt"
+        self.snap_device = "/dev/elastio-snap{}".format(self.minor)
+        self.snap_bkp = "./base-image.img"
```
Comment (e-kov) [inline tests/test_update_image.py:24]: This image is base only once after `setup-snapshot`. Then it's base+incremental, base+2incrementals etc ) Let's call it `backup.img` or somehow else.

File: tests/test_update_image.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+        loop_dev = util.loop_create(self.snap_bkp)
+        self.addCleanup(util.loop_destroy, loop_dev)
+
+        # Need to repair the xfs (see #63)
+        if self.fs == 'xfs':
+            util.mount(loop_dev, temp_dir, opts="nouuid")
+            util.unmount(temp_dir)
+
+        # For some reason, even valid xfs file system is showed
+        # as 'not valid' with xfs_repair v4.9.0
```
Comment (e-kov) [inline tests/test_update_image.py:69]: ...is shown as 'not valid'?

File: tests/test_update_image.py (hunk 2 above)
Comment (skypodolsky) [reply tests/test_update_image.py]: Done.

File: tests/test_update_image.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+import unittest
+
+import elastio_snap
+import util
+from devicetestcase import DeviceTestCase
+
+
+class TestUpdateImage(DeviceTestCase):
+    def setUp(self):
+        self.snap_mount = "/mnt"
```
[the remark this answers is not in the document]
Comment (skypodolsky) [reply tests/test_update_image.py]: Done.

File: tests/test_update_image.py (hunk 3 above)
Comment (skypodolsky) [reply tests/test_update_image.py]: Done.

File: tests/test_update_image.py (hunk 4 above)
Comment (skypodolsky) [reply tests/test_update_image.py:67]: Yes, `xfs_repair -n` simply returns '1' whatever the actual state of the filesystem is. Filesystem check simply doesn't work on that version.

File: tests/test_update_image.py [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+
+        # Need to repair the xfs (see #63)
+        if self.fs == 'xfs':
+            util.mount(loop_dev, temp_dir, opts="nouuid")
+            util.unmount(temp_dir)
+
+        # For some reason, even valid xfs file system is showed
+        # as 'not valid' with xfs_repair v4.9.0
+        if util.xfs_repair_version() != '4.9.0':
+            util.fsck(loop_dev, self.fs)
```
[the remark this answers is not in the document]
Comment (skypodolsky) [reply tests/test_update_image.py:69]: > What distro version has this xfs_repair 4.9.0?

Debian 9

> Does it mean that older xfs_repair versions are not affected by this bug?

We also have v4.5.0, it wasn't seen there

> And why there is no return code check?

It's done underneath with `subprocess.check_call()`. If it fails, the test fails as well. The same as we did for `mkfs` or dd``, for example.

File: tests/test_update_image.py [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+        self.addCleanup(elastio_snap.destroy, self.minor)
+
+        # preparing base image
+        util.dd(self.snap_device, self.snap_bkp, self.size_mb, bs="1M")
+        self.addCleanup(os.remove, self.snap_bkp)
+        write_testfile = "{}/{}".format(self.mount, file_name)
+
+        for i in range(1, iterations):
+            with open(write_testfile, "a") as f:
+                f.write("Attempt to destroy humanity #{}\n".format(i))
```
[the remark this answers is not in the document]
Comment (skypodolsky) [reply tests/test_update_image.py]: Okay, why not 🙂

File: tests/test_update_image.py (hunk 4 above)
Comment (skypodolsky) [reply tests/test_update_image.py]: Ah, to replace `showed` with `shown`)) Done!

File: tests/test_update_image.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (e-kov) [reply tests/test_update_image.py]: =)

File: tests/test_update_image.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (e-kov) [reply tests/test_update_image.py:69]: Ok. I see. Thanks. Interesting that it's a bug of a single version.

File: tests/test_update_image.py (hunk 4 above)
Comment (e-kov) [reply tests/test_update_image.py]: Yes, maybe "shown" sounds better in this context. I'm not an expert in this aria ) Thanks.

File: tests/devicetestcase.py (hunk 1 above)
Comment (e-kov) [reply tests/devicetestcase.py:33]: Makes sense. Ok.

File: .github/workflows/ci.yml [hunk 8]
```diff
@@ -181,7 +181,6 @@ jobs:
       - name: Install kernel module
         run: |
           vagrant ssh ${{env.INSTANCE_NAME}} -c 'sudo make install'
-          vagrant ssh ${{env.INSTANCE_NAME}} -c 'sudo make uninstall'
```
[the remark this answers is not in the document]
Comment (e-kov) [reply .github/workflows/ci.yml:184]: >  We can make it work with dirty hacks (like export LD_LIBRARY_PATH=... from python or anything similar)

Agree, it's not worth it. I just wonder how other people (not you and me) will figure out to do installation before the tests? That's it.
I've checked out your branch and tried `sudo ./elio-test.sh -t test_update_image` w/o `make install` in advance. And the error is 
```
subprocess.CalledProcessError: Command '['../utils/update-img', '/dev/elastio-snap5', '/tmp/elastio-snap_005/cow0', './backup.img']' returned non-zero exit status 127.
```
It's not obvious. 127 is "No such file or directory". But `../utils/update-img` exists in my case. And it returns `error while loading shared libraries: libelastio-snap.so.1: cannot open shared object file: No such file or directory`. Maybe we can print an error from it to the test's output. What do you think?

> Why do you think that not uninstalling the binaries would be the problem, and why do we uninstall them at all?

Just for the trivial test that `make uninstall` returns 0 and it's not broken. Once we had `make install` broken.

Let's add a final CI step after the tests before  `Detach external drives` or `Destroy a box` with `make uninstall` command.

File: tests/util.py [hunk 9]
```diff
@@ -122,6 +121,32 @@ def mkfs(device, fs="ext4"):
...
+    if fs == 'xfs':
+        cmd = ["xfs_repair", "-n", "-v", image]
+    else:
+        cmd = ["fsck." + fs, "-n", "-f", "-v", image]
+
+    subprocess.check_call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=10)
+
+def update_img(device, cow_file, bkp):
+    cmd = ["../utils/update-img", device, cow_file, bkp]
+    subprocess.check_call(cmd, stdout=subprocess.DEVNULL, stderr=subprocess.DEVNULL, timeout=180)
```
Comment (e-kov) [inline tests/util.py:139]: Well, there is an easy way to warn user about library absence at least by removing `stderr=subprocess.DEVNULL`. In this case user will see an error
```
../utils/update-img: error while loading shared libraries: libelastio-snap.so.1: cannot open shared object file: No such file or directory
```
What do you think?

Comment (skypodolsky) [reply tests/util.py:139]: I agree with that.

**Pull request merged**